book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,218 @@
+`is_ancestor(S, S_d)` is true by the definition of `S_d`.
+
+Now we show the protocol is sufficient to determine `is_ancestor(L_i, S)`.
+Because `L_i + N >= S` in this variant, and `N_Ancestors` has length `N`, then
+if `L_i`, is an ancestor, it has to be a member of `N_Ancestors`, so the
+protocol is sufficient.
+
+#### Variant 2: `L_i` < `N_Oldest`:
+
+##### Protocol:
+
+Search the vote state of the validator in `S_d` for `L_i`. If `L_i` is not
+found, assume it is not an ancestor.
+
+##### Proof of Correctness for Protocol:
+
+A vote for a slot will only be placed into the validator's vote state in a bank
+descended from the slot the vote is for. Therefore, a vote will only be found
+in a vote state if the vote is for an ancestor of the slot the vote state is
+found in. This protocol may falsely determine that `L_i` is not an ancestor of
+`S_d` if the vote was never accepted by a leader, but the result will be that
+the validator will take the more cautious approach of waiting for the lockout
+of the vote to expire, so it will not be slashed due to this failure.
+
+### Case 2: Calculating `is_ancestor(L_i, S_d)` when `L_i` >= `S`:
+
+##### Protocol: 
+
+If `S_d < L_i`, return false, because an ancestor cannot have a greater slot
+number. Otherwise, Let the bank state for slot `S_d` be called `B_d`. Check
+`B_d.ancestors().contains(L_i)`.
+
+##### Proof of Correctness for Protocol:
+
+**Lemma 1:** Given any `BankForks` and its root `R`, the bank state `B_d` for
+some slot `S_d`, where `B_d` is present in `BankForks.banks`, `B_d.ancestors()`
+must include all ancestors of `B_d` that are `>= R`
+
+**Proof:** Let `R` be the latest root bank in `BankForks`. The lemma holds at
+boot time because `R == S` and by step 4 of the "Boot Procedure", BankForks
+will only contain descendants of `S`,so each descendants' `ancestors` will only
+contain ancestors `>= R`. Going forward, by construction, ReplayStage only adds
+banks to `BankForks` if all of its ancestors are present and frozen. Thus
+because `B_d` is present in `BankForks.banks`, `B_d.ancestors()` must include
+all frozen ancestors `>= R`. Furthermore, `BankForks` prunes ancestors in its
+set of `banks` that are `< R'` after setting a new root `R'`, so this invariant
+is always true.
+
+**Lemma 2:** Given a root bank `R` of `BankForks` and some `L_i` in locktower,
+if `L_i >= S`, then `L_i >= R`.
+
+**Proof:** When we boot from a snapshot, we set `R == S`, and in this case
+because we are assuming `L_i > S`, then we know `L_i >= R` at bootup. Then when
+we set a new root in BankForks `R'` after booting, we guarantee this only
+happens if the locktower root is also set to `R'`. By the construction of
+locktower, all votes in the tower must be greater than the locktower root, so
+`L_i >= R'`. 
+
+For `Case 2` we assumed `L_i` >= `S`, so from Lemma 2 we know `L_i` >= `R`.
+Then from Lemma 1 we know its sufficient to check
+`B_d.ancestors().contains(L_i)`.
+
+
+## Boot Procedure:
```

Comment:
Some of the lemma's refer to this Boot Procedure (implies this section should be before the lemma's), should we move this section to right after `Snapshot Verification Overview`?